### DISCUSSION TITLE
What do all the parameters of `CACHE::update_replacement_state` means for replacement policies?

### DISCUSSION THREAD

**Discussion by Setu-Gupta** [ChampSim/ChampSim#438] [Q&A]
I'm trying to track how individual cache lines are flowing through the cache hierarchy. To do this, I added various print statements in my replacement policy. However, when I do this I see weird results such as cache lines getting evicted without being fetched in the first place. I expect to see a miss for a cache line before seeing its eviction as the miss would be the access that brings cache line in. The other alternative is that the prefetcher brought the line in. So to cover that case, I also track the prefetches being made.

To figure out what all cache requests I'm getting I'm using the various arguments of `CACHE::update_replacement_state`. I'm assuming that these arguments have the following meaning:
* `triggering_cpu`: This is the core with which the cache is associated.
* `set`: This is the cache set that is being accessed.
* `set`: This is the cache way that is being accessed.
* `full_addr`: This is the memory address which is being accessed.
* `ip`: This is the instruction pointer of the instruction which generates the memory address.
* `victim_addr`: This is the address of the cache line which is being evicted. If no cache line is being evicted, then this is 0.
* `type`: This is the type of access.
* `hit`: This is a boolean flag which tells if there was a hit.

I assume that `CACHE::update_replacement_state` is called on every cache access regardless of whether it is a hit, miss, or a fill.

Under these assumptions, I classify various cache requests as follows:
* `hit` and `access_type{type} == access_type::PREFETCH`: This implies that a prefetch request for `full_addr` was made which hit in the cache, i.e. this was a redundant prefetch request.
* `hit` and `access_type{type} != access_type::PREFETCH`: This implies that a demand request for `full_addr` was made which hit in the cache, i.e. demand hit
* `!hit` and `access_type{type} == access_type::PREFETCH`: This implies that a prefetch request for `full_addr` was made which missed in the cache. As a result, the cache generated a memory request for `full_addr` which resulted in the eviction of `victim_addr` due to conflicts.
* `!hit` and `access_type{type} != access_type::PREFETCH`: This implies that a demand request for `full_addr` was made which missed in the cache. As a result, the cache generated a memory request for `full_addr` which resulted in the eviction of `victim_addr` due to conflicts.
* `hit` and `victim_addr`: This should never happen.
* `!hit` and `!victim_addr`: This should never happen.

Are these assumptions and explanations correct?

**Comment by ngober** [comment]
The documentation page you are looking for is [here](https://champsim.github.io/ChampSim/master/Modules.html#replacement-policies). (Note that I think the develop version is better at the moment. I'm doing a lot of formatting improvements lately.)

In particular, you are incorrect as to the timing of the calls. `update_replacement_state()` is called on a hit or a fill, not on a miss. Other than that, I think your assessment of the classifications are correct.

As a side note, you should look at #434. It adds a hook for the replacement policy to observe fills separately from hits and misses. If you like it, let me know and we can merge it.

**Reply by Setu-Gupta** [reply]
Hi Nathan

Thanks for the reply. If I understand you correctly, the classifications that I came up with are correct. If that's the case then I should not be seeing evictions of cache lines before they are demanded. Although I may not see a cache fill, since as you mentioned `update_replacement_state()` is not called on fills, I should at least see a demand miss before an eviction.

However, this does not happen. To replicate this issue, you can have a look at my ChampSim repository which is fork of this repository [here](https://github.com/Setu-Gupta/ChampSim_SOTA). After cloning the repository, checkout the `cache_testing` branch. Then to replicate the issue, run the following commands after setting up ChampSim
```
$ ./config.sh SOTA_configs/tests/cache_test.json
$ make -j
$ ./bin/champsim_test --warmup_instructions 2000000 --simulation_instructions 5000000 path/to/400.perlbench-41B.champsimtrace.xz
```

If you examine the output, you'll see that many cache lines are evicted before they have a demand miss. I've disabled prefetchers for this configuration to narrow down the bug. In particular, look at `0x22c71f0`. This line is evicted twice from the L1D without having a demand miss even once.

**Reply by ngober** [reply]
I'm responding immediately to clarify, and I haven't checked out your work yet. But, I think you're still incorrect in your assessment of when the hook is called. The hook is not called on misses. It is called on hits and on fills. So, when the cache experiences a miss, the replacement policy is not notified until the fill occurs. (This distinction is my purpose for #434.) 

I would expect to see a fill before an eviction. If that isn't the case, maybe an inconvenient garbage value is being passed when an invalid way is filled. We could dig into that more precisely if we need to.
